src/networks/infinitesquarenetwork.jl

```diff
@@ -1,60 +1,77 @@
 """
-    InfiniteSquareNetwork{T,N}
+    InfiniteSquareNetwork{O}
 
-Abstract infinite tensor network consisting of a translationally invariant unit cell
-on a square lattice.
+Contractible square network. Wraps a matrix of 'rank-4-tensor-like' objects.
 """
-abstract type InfiniteSquareNetwork{T,N} end
+struct InfiniteSquareNetwork{O}
+    A::Matrix{O}
+    InfiniteSquareNetwork{O}(A::Matrix{O}) where {O} = new{O}(A)
+    function InfiniteSquareNetwork(A::Array{O,2}) where {O}
+        for (d, w) in Tuple.(CartesianIndices(A))
+            north_virtualspace(A[d, w]) ==
+            _elementwise_dual(south_virtualspace(A[_prev(d, end), w])) || throw(
+                SpaceMismatch("North virtual space at site $((d, w)) does not match.")
+            )
+            east_virtualspace(A[d, w]) ==
+            _elementwise_dual(west_virtualspace(A[d, _next(w, end)])) ||
+                throw(SpaceMismatch("East virtual space at site $((d, w)) does not match."))
+        end
+        return new{O}(A)
+    end
+end
```

Comment:
This is the concrete contractible network type discussed in #115 . It wraps a matrix of local rank-4-like objects, and just requires some access to the virtual spaces in all directions, which can be `ProductSpace`s.

Just adding a converter and a corresponding `rrule` for the physics network makes it so all contractions directly work with these objects, where only the contractions have to actually specialize on the local object type.